Add GMAC device tree binding

stephanosio opened the pull request:
This PR adds the device tree binding for the Atmel SAM GMAC Ethernet and converts the existing compatible SoCs (SAM E70 and V71) and boards (`sam_e70_xplained` and `sam_v71_xult`) to use it.

Includes commits from #23055.

Fixes #22997.

stephanosio committed:
dts: bindings: Add Atmel SAM-family GMAC binding

This commit adds the missing Atmel SAM-family GMAC Ethernet device
binding.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
dts: arm: atmel: Add SAM E70 GMAC instance

This commit adds a GMAC instance to the SAM E70 device tree, with the
chip revision-specific hardware queue count.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: sam_e70_xplained: Add revision B variant

This commit adds a new `sam_e70b_xplained` variant with the chip
revision B device tree.

In addition, this commit enables the newly added GMAC device tree
instance.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
dts: arm: atmel: Add SAM V71 GMAC instance

This commit adds a GMAC instance to the SAM V71 device tree, with the
chip revision-specific hardware queue count.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: sam_v71_xult: Add revision B variant

This commit adds a new `sam_v71_xult` variant with the chip revision B
device tree.

In addition, this commit enables the newly added GMAC device tree
instance.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
cc @nandojve

pfalcon reviewed (commented):
Just a comment to make this not show in my pending review list, as I'm not authority on DT stuff.

stephanosio commented:
@ulfalizer @galak Any suggested alternative for `dt_int_val`, which is apparently deprecated?

stephanosio committed:
CODEOWNERS: Add SAM E70 code owner

Add @nandojve, the SAM V71 platform maintainer, as a code owner for
the SAM E70 platform, as these two platforms share the same base and
are practically identical.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
@nandojve I am adding you as the SAM E70 platform code owner, since there currently is none and the `Codeowners` check is complaining about it.

stephanosio commented:
We are currently seeing USB-related build errors for SAMV71B because the ASF `usbhs.h` header in `hal_atmel` is missing some definitions for unknown reason.

https://github.com/zephyrproject-rtos/hal_atmel/blob/5a44c9d54dffd685fb6664a646922cfe41c5cf8e/asf/sam/include/samv71/component/usbhs.h#L1299-L1325

https://github.com/zephyrproject-rtos/hal_atmel/blob/5a44c9d54dffd685fb6664a646922cfe41c5cf8e/asf/sam/include/samv71b/component/usbhs.h#L1341-L1355

https://gist.github.com/stephanosio/cc7dd2049a740bdb13b1acdf6c6325af

stephanosio committed:
drivers: ethernet: sam_gmac: Use device tree values

This commit updates the Atmel SAM GMAC driver to use the device tree
values for GMAC hardware configuration.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Rebased on top of the revised #22961

stephanosio commented:
CI is currently failing due to https://github.com/zephyrproject-rtos/zephyr/issues/23415

mnkp reviewed (changes requested):
Since we introduced DTS node for GMAC we should remove usage of `ETH_SAM_GMAC_NAME`, `ETH_SAM_GMAC_IRQ_PRI`, `ETH_SAM_GMAC_MACx` Kconfig symbols as all the data are available via DTS.

stephanosio commented on dts/arm/atmel/same70q21b.dtsi:
```diff
@@ -0,0 +1,9 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <atmel/same70q21.dtsi>
```
I wanted to avoid creating duplicate code, but I agree that this approach is rather unconventional and can potentially be confusing.

I will just copy and paste as you suggested since the amount of duplicate code here isn't that much.

stephanosio commented:
> Since we introduced DTS node for GMAC we should remove usage of `ETH_SAM_GMAC_NAME`, `ETH_SAM_GMAC_IRQ_PRI`, `ETH_SAM_GMAC_MACx` Kconfig symbols as all the data are available via DTS.

@mnkp I have updated the GMAC driver to use the config values from the device tree.

stephanosio committed:
samples: net: gptp: Remove Kconfig MAC address for sam_e70_xplained

This commit removes the Kconfig MAC address configurations for the
`sam_e70_xplained` board, as these have been migrated to use the device
tree configurations.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

mnkp commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -2097,66 +2097,65 @@ static const struct ethernet_api eth_api = {
+static struct device DEVICE_NAME_GET(eth_sam_gmac);
 
-static void eth0_irq_config(void)
+static void eth_irq_config(void)
```
I believe it would be better to keep the old name `eth0_irq_config`. It's unlikely that we are ever going to have SAM series SoC with two ETH interfaces, still the function is instance specific and index 0 is matching `DT_INST_0_` defines from the DTS.

stephanosio replied:
Ok, keeping the `eth0_` prefix sounds reasonable, given that these are indeed instance-specific.

mnkp commented on dts/arm/atmel/same70b.dtsi:
```diff
@@ -0,0 +1,13 @@
+	interrupts = <39 0>, <66 0>, <67 0>, <71 0>, <72 0>, <73 0>;
+	interrupt-names = "gmac", "q1", "q2", "q3", "q4", "q5";
+	queue-num = <6>;
+};
```
@galak could you double check if this syntax is OK or we should prefer
```
/ {
	soc {
		gmac: ethernet@40050088 {
			...
		};
	};
};
```

stephanosio replied:
They mean two different things: what you suggested above instantiates a new device named "gmac", while that code sets properties for the already-instantiated device named "gmac".

stephanosio replied:
Note that the device `gmac` was already instantiated by `same70.dtsi` at this point.

galak replied:
I'm fine w/the syntax choice.  No real preference on how to do this in the soc dtsi files.  Leave it to the soc maintainer to decide what style to use.

stephanosio commented:
Rebased

stephanosio commented:
Rebased

nandojve commented:
Tested on SAMV71 with samples/net/telnet using net->ping and net->dns.

galak commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -0,0 +1,17 @@
+    reg:
+      required: true
+
+    queue-num:
```
Would be better as `num-queues`.  Is you have it, makes me think the property is to select a queue number to use.

stephanosio replied:
Used the name `queue-num` here to be consistent with the naming convention used by the GMAC driver, but sure, this can be somewhat misleading.

Will change it to `num-queues`.

galak commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -0,0 +1,17 @@
+    queue-num:
+      type: int
+      required: true
+      description: Number of hardware TX and RX queues
```
Just to be clear is the number of TX + RX queues?  (I'm assuming that there are the same number of RX & TX queues)

stephanosio replied:
Yes, this is the total number of hardware queues available (including TX and RX).

galak commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -31,14 +31,7 @@
-#else
-#define GMAC_QUEUE_NUM                  1
-#endif
+#define GMAC_QUEUE_NUM                  DT_INST_0_ATMEL_SAM_GMAC_QUEUE_NUM
```
Please use new DT_INST macros.  DT_INST_PROP(0, queue_num)

stephanosio replied:
Done

galak commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -2139,38 +2139,44 @@ static struct device DEVICE_NAME_GET(eth0_sam_gmac);
 
 static void eth0_irq_config(void)
 {
-	IRQ_CONNECT(GMAC_IRQn, CONFIG_ETH_SAM_GMAC_IRQ_PRI, queue0_isr,
```
Convert this to use new DT_INST macro syntax.

stephanosio replied:
Done

galak reviewed (changes requested):
If you want to pull out the commits for creating the 'B' variants that PR could get merged while the GMAC changes are still being done.

Please convert to use the new DT macro's in include/devicetree.h.

mnkp reviewed (approved):
LGTM, comments from @galak still need to be addressed.

stephanosio committed:
drivers: ethernet: sam_gmac: Convert to new DT_INST macros

This commit converts the Atmel SAM GMAC driver to use the new DT_INST
macro APIs.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
> If you want to pull out the commits for creating the 'B' variants that PR could get merged while the GMAC changes are still being done.
> 
> Please convert to use the new DT macro's in include/devicetree.h.

@galak I have added a new commit to convert the Atmel SAM GMAC driver to use the new DT_INST macros.

galak commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -10,6 +10,8 @@
 #ifndef ZEPHYR_DRIVERS_ETHERNET_ETH_SAM_GMAC_PRIV_H_
 #define ZEPHYR_DRIVERS_ETHERNET_ETH_SAM_GMAC_PRIV_H_
 
+#define DT_DRV_COMPAT atmel_sam_gmac
```
Move this from the header into the top of the driver source file. (before any includes).

stephanosio replied:
Done.

FYI, `eth_sam_gmac_priv.h` uses `DT_INST_PROP` macro, and we will be seeing very strange behaviours if this header gets included by any other file (though, since this is a "private header," it shouldn't really matter).

galak reviewed (commented):
in general looks good, one change associated with where the `#define DT_DRV_COMPAT` is done.

stephanosio commented:
> in general looks good, one change associated with where the `#define DT_DRV_COMPAT` is done.

@galak Fixed
